app/uk/gov/hmrc/exports/migrations/changelogs/cache/AddSequenceIdsToPackageInformation.scala

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.migrations.changelogs.cache
+
+import com.mongodb.client.MongoDatabase
+import org.bson.Document
+import org.mongodb.scala.model.Filters._
+import play.api.Logging
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+import uk.gov.hmrc.exports.migrations.changelogs.{MigrationDefinition, MigrationInformation}
+import uk.gov.hmrc.exports.migrations.TimeUtils.jsonWriter
+import uk.gov.hmrc.exports.models.DeclarationType.DeclarationType
+import uk.gov.hmrc.exports.models.declaration.AdditionalDeclarationType.AdditionalDeclarationType
+import uk.gov.hmrc.exports.models.declaration.DeclarationMeta.{sequenceIdPlaceholder, PackageInformationKey}
+import uk.gov.hmrc.exports.models.declaration.ExportsDeclaration.Mongo
+import uk.gov.hmrc.exports.models.declaration._
+
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
+
+class AddSequenceIdsToPackageInformation extends MigrationDefinition with Logging {
+
+  override val migrationInformation: MigrationInformation =
+    MigrationInformation(
+      id = s"CEDS-4445 Add a 'sequenceId' field to 'PackageInformation' sub-documents",
+      order = 20,
+      author = "Lucio Biondi",
+      // Can be removed (default is false) once this migration is deployed to all envs (production & integration included)
+      runAlways = true
+    )
+
+  override def migrationFunction(db: MongoDatabase): Unit = {
+    logger.info(s"Applying '${migrationInformation.id}' db migration...")
+
+    val collection = db.getCollection("declarations")
+
+    val batchSize = 100
+
+    val filter =
+      elemMatch("items", elemMatch("packageInformation", not(exists("sequenceId"))))
+
+    val result = collection
+      .find(filter)
+      .batchSize(batchSize)
+      .asScala
+      .foldLeft(0) { (counter, document) =>
+        Try(Json.parse(document.toJson(jsonWriter)).as[ExportsDeclaration](AddSequenceIdsToPackageInformation.reads)) match {
+          case Failure(exc) =>
+            val id = document.get("_id")
+            val error = exc.getMessage
+            logger.error(s"Error parsing document with _id($id): $error")
+            counter + 1
```

Comment:
Just occurred to me that these failures are actually about failure to read the json documents using these custom 'former' reader functions. If this migration is ever run in the future and records match the given filter, we might get read failures because the structure of the entities has since been altered :thinking:

Reply:
Let's postpone a discussion on this topic to the next dev catchup, as I have something to say on this, OK?